### PR TITLE
GVF updates

### PR DESCRIPTION
**Pull request by PakhiAgarwal**

### CODE REVIEW

File: generateBins.py [hunk 1]
```diff
@@ -1,86 +1,47 @@
...
-#%matplotlib inline
 
 
 # Import our libraries
 sys.path.append(os.path.join(os.path.dirname(__file__), "include"))
 
-from ascFile import *
-from calibrationLib import *
+from include.ascFile import *
+from include.calibrationLib import *
```
Comment (rjzupkoii) [inline generateBins.py:14]: Still not sure why there is a regression here.

File: generateBins.py [hunk 2]
```diff
@@ -1,86 +1,47 @@
...
 
 
 # TODO Still need a good way of supplying these
 PFPR_FILE       = "rwa_pfpr2to10.asc"
 POPULATION_FILE = "rwa_population.asc"
-
+TREATMENT_FILE = "rwa_treatment.asc"
```
Comment (rjzupkoii) [inline generateBins.py:20]: This shouldn't be needed anywhere.

File: generateBins.py [hunk 3]
```diff
@@ -1,86 +1,47 @@
...
-    return gvf
-
-
 def process(configuration, gisPath = ""):
     # Load the configuration
     cfg = load_configuration(configuration)
 
+    # getting data ready for binning
+    # data should be 1-dimensional array, python list or iterable
+    myArray = np.loadtxt(POPULATION_FILE, skiprows=6)
```
Comment (rjzupkoii) [inline generateBins.py:32]: Use the include/ascFile.py script for this, that's what it's there for.

File: generateBins.py [hunk 4]
```diff
@@ -93,6 +54,22 @@ def process(configuration, gisPath = ""):
...
+
+        # GVF Implementation
+        gvf = 0.0
+        nclasses = 20
+        while gvf < 0.8:
+            gvf_t = goodness_of_variance_fit(treat_1d, nclasses)
+            nclasses += 1
+        print("The value of Goodness of Variance fit is:", gvf)
+        if (gvf < 0.7):
+            print("Warning: GVF too low")
```
Comment (rjzupkoii) [inline generateBins.py:70]: Don't worry about the treatments for now. It's not needed for RWA and for BFA we used a different approach.

Comment (rjzupkoii) [inline generateBins.py:70]: Avoid copy-paste coding as well, using this block of code more than once is a sign that it should be a function.

File: generateBins.py [hunk 5]
```diff
@@ -1,86 +1,47 @@
...
     # TODO Add the stuff for the population bins!
+    # GVF Implementation
+    gvf = 0.0
+    nclasses = 20
+    while gvf < 0.8:
+        gvf = goodness_of_variance_fit(array_1d, nclasses)
+        nclasses += 1
+    print("The value of Goodness of Variance fit is:", gvf)
+    if (gvf < 0.7):
+        print("Warning: GVF too low")
```
Comment (rjzupkoii) [inline generateBins.py:44]: Okay, from the GVF we should know how to bin the population data. However, I don't seem to see that taking place.

File: include/calibrationLib.py [hunk 6]
```diff
@@ -1,10 +1,13 @@
 # calibrationLib.py
-#
+#GVF Source: https://medium.com/analytics-vidhya/jenks-natural-breaks-best-range-finder-algorithm-8d1907192051
```
Comment (rjzupkoii) [inline include/calibrationLib.py:2]: Reusing code is _very_ common in the programming community, the same block appeared back in 2015 on CrossValidated: https://stats.stackexchange.com/a/144075

File: include/calibrationLib.py [hunk 7]
```diff
@@ -1,10 +1,13 @@
...
+
 # This module includes functions that are intended for use with calibration functions.
 import csv
 import os
 import yaml
-
+from jenkspy import jenks
+import jenkspy
+import numpy as np
```
Comment (rjzupkoii) [inline include/calibrationLib.py:10]: Added NumPy dependency to README.

**Review by rjzupkoii — CHANGES_REQUESTED** [review]

File: generateBins.py (hunk 5 above)
Comment (PakhiAgarwal) [reply generateBins.py:44]: I think that should be the classes. As we are taking number of classes as 20 (based on hit and trial; we can increment and decrement the number of classes to get the value for perfect fitting number; ~1) and the GVF is greater than .95, which makes the fit as ~Perfect fit. The value of the break points generated using Jenks is stored in classes, so I think the value in classes must determine the values of bins.

Comment (rjzupkoii) [reply generateBins.py:44]: @PakhiAgarwal Right, off hand it seems like 20 is likely a high bar for the data sets I've seen so far so I'd start the looking at a low value like 5 or so and then proceed from there. Also, I just noticed this, but there's the possibility for an infinite loop in the code right now.

Comment (PakhiAgarwal) [reply generateBins.py:44]: Okay. Based on the data, I'll start from 5 upto 20-25 until I get some number approximate to ~1. This will also help to check for infinite loop.

Comment (PakhiAgarwal) [reply generateBins.py:44]: @rjzupkoii These are the values I got:

    `5            0.8187998573940524
     6            0.8400063031786603
     7            0.8725093908883502
     8            0.9105305546827702
     9            0.9206628960735237
     10           0.9282909873982712
     11           0.9410801996635164
     12           0.9453292045975631
     13           0.9460764761434216
     14           0.9560381166408211
     15           0.9565395731869651
     16           0.956969066379069
     17           0.9574810318860014
     18           0.9592570348935402
     19           0.9594522531572597
     20           0.9596282367471943
     21           0.9597394491506118
     22           0.9598343385619622
     23           0.9599260836086719
     24           0.9599659577944439
     25           0.9600349654202794`
    
Looking at these values I think 14 or 15 could be a good value.

File: generateBins.py [hunk 8]
```diff
@@ -28,59 +18,28 @@
...
-
-
 def process(configuration, gisPath = ""):
     # Load the configuration
     cfg = load_configuration(configuration)
+    filename = os.path.join(gisPath, PFPR_FILE)
+
+    # getting data ready for binning
+    # data should be 1-dimensional array, python list or iterable
+    myArray = np.loadtxt(load_asc(filename), skiprows=6)
```
Comment (rjzupkoii) [inline generateBins.py:32]: `load_asc` returns the ASC data as floating point data, so I'm not sure why you need to do this.

File: generateBins.py [hunk 9]
```diff
@@ -28,59 +18,28 @@
...
+    array_1d = myArray.flatten()
 
     # TODO Add the stuff for the population bins!
+    # GVF Implementation
+
+    gvf = goodness_of_variance_fit(array_1d, 14)
+    print("The value of Goodness of Variance fit is:", gvf)
+    if (gvf < 0.7):
+        print("Warning: GVF too low")
```
Comment (rjzupkoii) [inline generateBins.py:41]: Alright, on the right track, but think of something more along the lines of `find_best_bins(data, minimum, maximum)` that returns the best number of bins from that range and the GVF. Then you can point the GVF if need be and call another scripts like `get_breaks(data, number_of_bins)` to get the breaks that are used in place of `POPULATION_BINS`.

**Review by rjzupkoii — CHANGES_REQUESTED** [review]

File: generateBins.py (hunk 5 above)
Comment (rjzupkoii) [reply generateBins.py]: Some of this should be starting to look familiar from the machine learning courses (e.g., iterative optimization algorithms like gradient descent) you have taken since that's effectively what we are doing here. So there's a couple different ways that we can process the values. 

1. Use best GVF value that we got (i.e., 25); however, we have to be careful about overfitting the data and being mindful of our business rules (i.e., minimize the number of bins so that calibration runs faster). So we can rule that approach out.
2. Use the smallest number of bins with a GVF greater than a given value. Since we are looking for GVF > 0.8 it looks like 5 would have been acceptable. However, based upon the knowledge of the data we are working with, 5 bins might calibrate fast, but depending on where the breaks are they could give us some odd results. I'd prefer a few more bins if possible.
3. Use the smallest number of bins returned at the point where the GVF ceases to change in a a meaningful way. Looking at the code it appears that the hundredths place is a good marker for this and if a `abs(gvf(a) - gvf(b)) <= 0.01` we might have found a result that is "good enough". For this dataset that condition would be satisfied when the number of bins goes from 9 to 10, suggesting that either of those values could work. Incidentally 10 bins were also originally fit for the dataset.

Comment (PakhiAgarwal) [reply generateBins.py]: That's what I was thinking, to use some ML technique. From 5 to 8 there is a drastic difference in value, then from 8 to 11 we can see the similar difference. But from 14 to 24 it remains almost the same, with little changes. I think we can use some algorithm to find the most optimal value. In this case, I think we can take 14 as most optimal because between 14 to 24, it remains almost the same with little difference. 
I'll look into it more. But for now maybe we can use this.

Comment (rjzupkoii) [reply generateBins.py]: @PakhiAgarwal This entire exercise is an application of ML techniques, so we are good there. :) Don't forget that we really do need to avoid overfitting (i.e., too many bins) with these datasets.

File: include/ascFile.py [hunk 10]
```diff
@@ -89,3 +89,17 @@ def write_asc(ascheader, ascdata, filename):
...
+# getting data ready for binning
+def bin_asc(filename):
+
+    # data should be 1-dimensional array, python list or iterable
+
+    with open(filename) as ascfile:
+        myArray = np.loadtxt(ascfile, skiprows=6)
+
+    array_1d = myArray.flatten()
+    return array_1d
```
Comment (rjzupkoii) [inline include/ascFile.py:103]: I see where you are going with this, but this really should be in something like `calibrationLib` instead. Also, still not sure why you need to use `loadtxt` to get the data.

File: generateBins.py [hunk 11]
```diff
@@ -27,15 +27,10 @@ def process(configuration, gisPath = ""):
...
-    # getting data ready for binning
-    # data should be 1-dimensional array, python list or iterable
-    myArray = np.loadtxt(load_asc(filename), skiprows=6)
-    array_1d = myArray.flatten()
-
     # TODO Add the stuff for the population bins!
     # GVF Implementation
-
-    gvf = goodness_of_variance_fit(array_1d, 14)
+    b = bin_asc(filename)
```
Comment (rjzupkoii) [inline generateBins.py:32]: Avoid one character variable names outside of limited cases like loop indicies.

**Review by rjzupkoii — CHANGES_REQUESTED** [review]

File: include/ascFile.py (hunk 10 above)
Comment (PakhiAgarwal) [reply include/ascFile.py]: As I am using numpy, so I thought it would be better to convert it into numpy array and start processing the file. Let me know if doesn't look good.

Comment (rjzupkoii) [reply include/ascFile.py]: That's not unreasonable, but besides the `flatten` function where all are you using NumPy?

Comment (PakhiAgarwal) [reply include/ascFile.py]: For now its used only for `flatten`

Comment (rjzupkoii) [reply include/ascFile.py]: @PakhiAgarwal You can use an expression like this to do the same operation then: `array = list(i for j in matrix for i in j)`

Comment (PakhiAgarwal) [reply include/ascFile.py]: Yeah. We can use this expression, but I think using numpy will have same effect; the complexity of both is comparable. 
Do we have to avoid using built in libraries?

Comment (rjzupkoii) [reply include/ascFile.py]: @PakhiAgarwal That's a good question and to keep things brief, in addition to the computational complexity, its also important to keep in mind the broader aspects of a project. A general rule of thumb is to avoid importing additional dependencies unless there is a compelling need for them since the dependencies place additional burdens on a project (e.g., make sure they are documented, installed, upgraded, etc.). There was [a case a couple years ago](https://arstechnica.com/information-technology/2016/03/rage-quit-coder-unpublished-17-lines-of-javascript-and-broke-the-internet/) where the removal of a JavaScript dependency caused cascading failures for awhile. Since this would be adding a dependency to do something that we can do ourselves with a single line of code, it doesn't make much sense to add that dependency. Conversely, if we were making heavy use of NumPy throughout it would make sense to use the `flatten` function for consistency and readablity.

Comment (PakhiAgarwal) [reply include/ascFile.py]: Sounds reasonable.

File: generateBins.py [hunk 12]
```diff
@@ -25,69 +13,46 @@
...
     cfg = load_configuration(configuration)
+    filename = os.path.join(gisPath, PFPR_FILE)
 
     # TODO Add the stuff for the population bins!
+    # GVF Implementation
+    binning = bin_asc(filename)
+    #gvf = goodness_of_variance_fit(binning, 14)
+    #print("The value of Goodness of Variance fit is:", gvf)
+    #if gvf < 0.7:
+    #    print("Warning: GVF too low")
```
Comment (rjzupkoii) [inline generateBins.py:35]: Remove instead of commenting out code

File: generateBins.py [hunk 13]
```diff
@@ -25,69 +13,46 @@
...
+
+    list_one = []
+    list_two = []
```
Comment (rjzupkoii) [inline generateBins.py:38]: There are better data structures for this and the variables should have clear names. Personally I'd use a `dictionary`.

File: generateBins.py [hunk 14]
```diff
@@ -25,69 +13,46 @@
...
+    for x in range(5, 26):
```
Comment (rjzupkoii) [inline generateBins.py:39]: It's not too bad in this instance, but it's good practice to avoid raw numbers and use constants instead. When they are used, you should have at least a comment indicating why this particular range.

File: generateBins.py [hunk 15]
```diff
@@ -25,69 +13,46 @@
...
+        GVF = goodness_of_variance_fit(binning, x)
+        if GVF < 0.7:
```
Comment (rjzupkoii) [inline generateBins.py:41]: Ditto above comment.

File: generateBins.py [hunk 16]
```diff
@@ -25,69 +13,46 @@
...
+            print("Warning: GVF too low")
+
+        if GVF >= 0.94:
```
Comment (rjzupkoii) [inline generateBins.py:44]: Ditto above comment, but also why is 0.94 special. Remember, that this script is going to be used for more than just Rwanda. There's also the Burkina Faso data we currently have, and in the future we will be looking at more countries.

**Review by rjzupkoii — CHANGES_REQUESTED** [review]

File: generateBins.py (hunk 16 above)
Comment (PakhiAgarwal) [reply generateBins.py:44]: Would be it okay if we use `0.9`? Will it be good for all the data we have?

File: generateBins.py (hunk 14 above)
Comment (PakhiAgarwal) [reply generateBins.py:39]: as these values are used only once in the whole code, so I haven't defined constants.

File: generateBins.py (hunk 16 above)
Comment (rjzupkoii) [reply generateBins.py:44]: @PakhiAgarwal Using a single value for all data is always very tricky, so some sort of heuristic is usually a much better route since they tend to be much more flexible. Recall the comment earlier about stopping the fitting when `abs(gvf(previous) - gvf(current)) <= 0.1`? That's likely a good spot to stop as a basic heuristic and also has the advantage when we a good fit can be found quickly we can go ahead and stop running. For a larger countries with more cells that can be a very valuable feature as well.

Comment (PakhiAgarwal) [reply generateBins.py:44]: I thought about using this, but in our current dataset the number of bins will be 7 or 8, this value will be even more less for other datasets maybe 4 or 5.
Would it be good for our work?

Comment (rjzupkoii) [reply generateBins.py:44]: @PakhiAgarwal Correct, for other data sets the value might be lower. You might want to use a range from 2 to 30 or so just to keep things simple.

File: generateBins.py [hunk 17]
```diff
@@ -25,69 +13,50 @@
...
+    for x in range(5, 30):
+        GVF = goodness_of_variance_fit(binning, x)
+        # GVF less than 0.7 might not be the best fit for data; GVF ~ 0 (No fit), GVF ~ 1 (Perfect fit)
+        if GVF < 0.7:
+            print ("Bins=", x,"GVF = ", GVF, "Warning: GVF too low")
+        # Storing the values of bins and GVF in a list
+        list_bins.append(x)
+        list_GVF.append(GVF)
+    # computing difference between GVF values and storing them as (key, value) pairs in dictionary
+    diff_GVF = {list_bins[i]: abs(list_GVF[i] - list_GVF[i - 1]) for i in range(len(list_bins))}
```
Comment (rjzupkoii) [inline generateBins.py:45]: You're on the right track here, but you can find the while going through each bin size in the range, then just break out of the for loop once the conditions is satisfied. Since we want to use the smallest number of bins, using the previous bin size should be fine. That also makes things a bit interesting if we reach the end of the loop without the heuristic being meet, it's up to you how to handle that.

File: generateBins.py [hunk 18]
```diff
@@ -25,69 +13,50 @@
...
+    for key, value in diff_GVF.items():
+        # taking such value of bin for which the difference between consecutive GVF values is greater than or equal to 0.1
+        if value >= 0.1:
+            print("Total number of bins", key)
+            # corresponding to the computed bin(key), finding the value of GVF
+            GVF = goodness_of_variance_fit(binning, key)
+            print("The value of Goodness of variance fit", GVF)
```
Comment (rjzupkoii) [inline generateBins.py:52]: Not quite sure where you are going here though - remember that once we know the number of bins with an acceptable GVF score (i.e., > 0.7 and with the heuristic being satisfied) then we want to use the bin values returned by Jenks in lieu of `POPULATION_BINS`.

**Review by rjzupkoii — CHANGES_REQUESTED** [review]

File: generateBins.py [hunk 19]
```diff
@@ -25,69 +13,53 @@
...
+    list_bins = []
+    list_GVF = []
+    X_MIN = 5
+    X_MAX = 30
+    # Assuming the range for bins to be between 5 and 30 (computing the optimal number of bins w.r.t. GVF), to avoid overfitting or underfitting of data
+    for x in range(X_MIN, X_MAX):
+        GVF, list_bins = goodness_of_variance_fit(binning, x)
+        # GVF less than 0.7 might not be the best fit for data; GVF ~ 0 (No fit), GVF ~ 1 (Perfect fit)
+        if GVF < 0.7:
+            print ("Bins=", x,"GVF = ", GVF, "Warning: GVF too low")
```
Comment (rjzupkoii) [inline generateBins.py:42]: No need to print the low GVF when you are still iterating through the bin sizes. A low GVF warning should be printed after the search space has been exhausted.

File: generateBins.py [hunk 20]
```diff
@@ -25,69 +13,53 @@
...
+        list_GVF.append(GVF)
+
+    # computing difference between GVF value
+    for i in range(len(list_GVF)):
+        # taking such value for which the difference between consecutive GVF values is greater than or equal to 0.1
+        if abs(list_GVF[i] - list_GVF[i - 1]) >= 0.1:
+            print("The value of Goodness of variance fit", list_GVF[i])
+            # corresponding to value of GVF, returning all bins
+            print("The value of bins", list_bins[:i + X_MIN])
+            break
```
Comment (rjzupkoii) [inline generateBins.py:54]: Again, you don't need to wait until all of the bin sizes have been checked, you can do this while you are iterating through them if you are applying a heuristic for the search.

File: include/calibrationLib.py [hunk 21]
```diff
@@ -251,4 +258,48 @@ def query_betas(connection, studyId, filterZero=True, filename="data/calibration
...
+    return gvf, classes
+
+def classify(value, breaks):
+    for i in range(1, len(breaks)):
+        if value < breaks[i]:
+            return i
+    return len(breaks) - 1
+
+# getting data ready for binning
+def bin_asc(filename):
```
Comment (rjzupkoii) [inline include/calibrationLib.py:297]: This is a poor function name since it implies that the result of the function will be a binned data set, not a data set that is ready for binning.

File: generateBins.py [hunk 22]
```diff
@@ -25,69 +13,53 @@
...
 
 # TODO Still need a good way of supplying these
-PFPR_FILE       = "rwa_pfpr2to10.asc"
+PFPR_FILE = "rwa_pfpr2to10.asc"
 POPULATION_FILE = "rwa_population.asc"
 
-
 # TODO Determine the bins computationally
 # Following bins are for Rwanda
 POPULATION_BINS = [2125, 5640, 8989, 12108, 15577, 20289, 27629, 49378, 95262, 286928]
```
Comment (rjzupkoii) [inline generateBins.py:21]: It seems like we are still missing the fine step of actually generating the population bin values using the GVF data.

**Review by rjzupkoii — CHANGES_REQUESTED** [review]

File: generateBins.py (hunk 20 above)
Comment (PakhiAgarwal) [reply generateBins.py]: It is returning only those bins for which the if condition is satisfied.

File: generateBins.py (hunk 22 above)
Comment (PakhiAgarwal) [reply generateBins.py:21]: pop_bins is generating the population bins

Comment (rjzupkoii) [reply generateBins.py:21]: @PakhiAgarwal If that's the case then you should go ahead and complete the rest of the task.

**Review by rjzupkoii — CHANGES_REQUESTED** [review]

File: loader.py [hunk 23]
```diff
@@ -1,286 +0,0 @@
-#!/usr/bin/python
-
-##
-# loader.py
```
Comment (rjzupkoii) [inline loader.py:4]: Why was this deleted? You are going to need to put the file back to merge the pull request since GitHub is logging a conflict between your code and this repository.

File: generateBins.py [hunk 24]
```diff
@@ -25,69 +13,56 @@
...
+    for x in range(X_MIN, X_MAX):
+        GVF, pop_bins = goodness_of_variance_fit(binning, x)
+        # Storing the values of bins and GVF in a list
+        pop_bins.append(x)
+        list_GVF.append(GVF)
+
+    # computing difference between GVF value
+    for i in range(len(list_GVF)):
+        # taking such value for which the difference between consecutive GVF values is greater than or equal to 0.1
+        if abs(list_GVF[i] - list_GVF[i - 1]) >= 0.1:
```
Comment (rjzupkoii) [inline generateBins.py:46]: Not `abs(gvf(a) - gvf(b)) <= 0.01` ?

**Review by rjzupkoii — CHANGES_REQUESTED** [review]

File: loader.py (hunk 23 above)
Comment (PakhiAgarwal) [reply loader.py:4]: I deleted by mistake

File: generateBins.py (hunk 24 above)
Comment (PakhiAgarwal) [reply generateBins.py:46]: Corrected

File: include/calibrationLib.py [hunk 25]
```diff
@@ -251,4 +258,51 @@ def query_betas(connection, studyId, filterZero=True, filename="data/calibration
...
+# getting data ready for binning
+def data_bin(filename):
+
+    # data should be 1-dimensional array, python list or iterable
+
+    with open(filename) as ascfile:
+        myArray = np.loadtxt(ascfile, skiprows=6)
+
+    myArray[myArray == -9999] = np.nan
+    myArray = myArray[~np.isnan(myArray)]
```
Comment (rjzupkoii) [inline include/calibrationLib.py:305]: Remember, you are working with ASC data which is a defined file type, which is why using a loader for the type is the preferred option since the header defines what the `nodata` value is. While `-9999` is being used as the `nodata` value here, it's entirely possible that the raster files for the next project will use a different value.

**Review by rjzupkoii — CHANGES_REQUESTED** [review]

File: include/calibrationLib.py (hunk 25 above)
Comment (PakhiAgarwal) [reply include/calibrationLib.py:305]: That's what I was thinking why we have `-9999` as `no data` value. Can you tell if `nodata` value for other files is always less than `0`.

Comment (PakhiAgarwal) [reply include/calibrationLib.py:305]: Also let me know what is the criteria for determining the value of `nodata`. Like for `populationBins.asc`, how is the value -9999 determined? Why is it only -9999 and not any other value for ex., -22, -30, -5000 ?

Comment (rjzupkoii) [reply include/calibrationLib.py:305]: @PakhiAgarwal There's no real way to know what it will be, but that's why the value is supplied in the header:

```python
[header, data] = load_asc(filename)
nodata = header['nodata']
```

Comment (PakhiAgarwal) [reply include/calibrationLib.py:305]: Okay, so it means `nodata` value can either be positive or negative. 
If it is positive then what happens if it clashes with other values, for ex., if `nodata` is say `20`, then can we have `20` as other actual `real value`?

Comment (rjzupkoii) [reply include/calibrationLib.py:305]: @PakhiAgarwal If the `nodata` value clashes with a real data value then it means the file is not correct and there are likely other problems that we would be worried about.
